### PR TITLE
Windows doesn't allow renaming files that are open (attempt 2)

### PR DESCRIPTION
**Pull request by Fydon**
This is a replacement for #121. All the tests now pass and the changes have been rebased to a recent version of the develop branch.

### CODE REVIEW

**Comment by cppforlife** [comment]
@Fydon will pull it in.

File: installation/tarball/provider_test.go [hunk 1]
```diff
@@ -77,18 +78,28 @@ var _ = Describe("Provider", func() {
...
 			Context("when tarball is not present in cache", func() {
 				var (
-					tempDownloadFilePath string
+					tempDownloadFilePath1 string
+					tempDownloadFilePath2 string
+					tempDownloadFilePath3 string
 				)
 
 				BeforeEach(func() {
-					tempDownloadFile, err := ioutil.TempFile("", "temp-download-file")
```
Comment (cppforlife) [inline installation/tarball/provider_test.go]: does this test fail on windows if not changed to have 3 temp files?

File: releasedir/fs_blobs_dir.go [hunk 2]
```diff
@@ -244,6 +244,8 @@ func (d FSBlobsDir) TrackBlob(path string, src io.ReadCloser) (Blob, error) {
 
 	blobs[idx] = Blob{Path: path, Size: fileInfo.Size(), SHA1: sha1}
 
+	tempFile.Close()
```
Comment (cppforlife) [inline releasedir/fs_blobs_dir.go]: any chance for a test for this one? nothing currently fails?

File: installation/tarball/provider_test.go (hunk 1 above)
Comment (Fydon) [inline installation/tarball/provider_test.go]: This test fails everywhere now without 3 temp files, as the file is being closed on each [retry][1]. The real file system is creating a new temp file on each of the 3 retries, but the test was previously only using one temp file for all retries.

[1]: https://github.com/cloudfoundry/bosh-cli/pull/171/files/f78baee58902f252afde22804b19549c0618b805#diff-860904024a4888f514f430f2bffc5c41R120

File: releasedir/fs_blobs_dir.go (hunk 2 above)
Comment (Fydon) [inline releasedir/fs_blobs_dir.go]: Sure, but what should that test do? This method does [defer the close of the temp file][1], but now the defer close will fail if this line is reached. The defer close isn't checking for failure, so this close is causing a silent failure, but also preventing an error in Windows environments.

[1]: https://github.com/cloudfoundry/bosh-cli/pull/171/files/f78baee58902f252afde22804b19549c0618b805#diff-91f1299e44f3b411467c957c91b729bbL208

**Comment by Fydon** [comment]
@cppforlife Should there be a defer close in `provider.go` to ensure that `downloadedFile` is always closed? In its current state could there be a resource leak?

Should I make defer close execute where I've added close? It seems that I could do this by splitting part of the function into another function or [using embedded functions][1]. The reason I ask this is that it might be confusing to have two close calls in the same function for the same resource, i.e. defer close and close.

[1]: https://play.golang.org/p/Xjy1g6W1Qf

**Pull request merged**